### PR TITLE
Make the corpus remote wrt workers

### PR DESCRIPTION
**Pull request by mtrofin**
This change removes the assumption that the corpus is co-located with the workers. The new design relies on the assumption that the workers are reachable via a high-throughput network (datacenter setup). The trainer prefetches the modules for the next iteration and sends them as data blobs to the workers.

The rest of the change trickles out from this:

- the corpus operates internally with a metadata object about modules, `CorpusElement`
- workers are given a `LoadedModuleSpec` which contains the data of the module (and its thinlto index, if needed).
- command line final shape is done on the worker side

This had the side-effect that responsibilities could be reassigned as follows:

- the corpus constructor handles preparing the `CorpusElement`s, with all the necessary flag modification handlers (i.e. add/delete/replace)
- flag modification logic is generic - it doesn't know anything about thinlto, '-cc1', etc

In turn, this impacted testing.

This change impacts purely local training similar to how a previous change remoting the policy does - local files are copied by workers redundantly. It doesn not appear to have any impact on training performance, but should we want to, we can optimize the local scenario under the hood.

### CODE REVIEW

File: compiler_opt/rl/compilation_runner.py [hunk 1]
```diff
@@ -340,6 +342,7 @@ def collect_data(
       ValueError if example under default policy and ml policy does not match.
     """
     with tempfile.TemporaryDirectory() as tempdir:
+      module_spec = loaded_module_spec.to_module_spec(tempdir)
```
[the remark this answers is not in the document]
Comment (yundiqian) [reply compiler_opt/rl/compilation_runner.py:345]: and this is going to do the file writing for local training as well

File: compiler_opt/rl/compilation_runner.py [hunk 2]
```diff
@@ -24,7 +24,7 @@
 import threading
 from typing import Dict, List, Optional, Tuple
 
-from absl import flags
+from absl import flags, logging
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply compiler_opt/rl/compilation_runner.py:27]: ack - do we want that for the line below, too - and we just do listing for typing? (i.e. `from typing import Dict, List, Tuple` the only instance this is ok)?

[the remark this answers is not in the document]
Comment (yundiqian) [reply compiler_opt/rl/compilation_runner.py:27]: I think so, for honoring google3 python coding style.

File: compiler_opt/rl/corpus.py [hunk 3]
```diff
@@ -32,11 +34,128 @@ class ModuleSpec:
...
+        orig_options=self.orig_options,
+        additional_flags=self.additional_flags,
+        delete_flags=self.delete_flags,
+        replace_flags=self.replace_flags)
+    return ModuleSpec(
+        name=self.name, exec_cmd=tuple(cmdline), size=len(self.module))
+
+
+@dataclass(frozen=True)
+class CorpusElement:
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply compiler_opt/rl/corpus.py:129]: It's in the intent. The `ModuleSpec` is basically a ready to use command line, everything has been prepared and ready to use. `CorpusElement` is tied to the `Corpus` object, it's very much not ready to use, the flag converters need to be run over it.

We could, actually, reduce the concept count to 1, and to avoid confusion I'll use this patch's terminology:
- remove `ModuleSpec` and replace with just the `Tuple[str, ...]` that's the fully resolved command line args
- `s/CorpusElement/ModuleSpec`

Basically, in the patch, the extra flags of ModuleSpec are not used, except for `name` which is used to raise an error. The error can be handled on the training side, who knows which corpus element this error pertains to (incl. to build up a repro script, out of the scope of this patch. But i'd just use the command line right now, with this proposal).

wdyt?

File: compiler_opt/rl/corpus.py [hunk 4]
```diff
@@ -32,11 +34,128 @@ class ModuleSpec:
...
+  """
+  name: str
+  module: bytes
+  thinlto_index: Optional[bytes] = None
+  orig_options: Tuple[str, ...] = ()
+  additional_flags: Tuple[str, ...] = ()
+  delete_flags: Tuple[str, ...] = ()
+  replace_flags: Optional[Dict[str, str]] = None
+
+  def _create_files_and_get_context(self, local_dir: str):
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply compiler_opt/rl/corpus.py:101]: It's not handling race conditions, intentionally. The caller is expected to create separate `local_dir`s (which it does). An alternative could be to let this function do that, but then it'd have to let the caller handle cleaning up (deleting the tempdir). In the end, the lifetime management of the directory is so much easier if left to the caller to fully handle in the first place.

Adding a comment though about the expectation from the caller. Adding it to `to_module_spec` because that's the public one.

File: compiler_opt/rl/compilation_runner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtrofin) [reply compiler_opt/rl/compilation_runner.py:345]: Yes. This is discussed in the patch message.

File: compiler_opt/rl/corpus.py [hunk 5]
```diff
@@ -104,192 +232,160 @@ def __init__(self,
...
+    sampled_specs = self._sampler(self._contents, k=k)
     if sort:
       sampled_specs.sort(key=lambda m: m.size, reverse=True)
     return sampled_specs
 
-  def filter(self, p: re.Pattern):
-    """Filters module specs, keeping those which match the provided pattern."""
-    return Corpus.from_module_specs(
-        (ms for ms in self._module_specs if p.match(ms.name)))
+  def load_corpus_element(self,
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply compiler_opt/rl/corpus.py:334]: We need a separation between metadata used by the corpus to organize files, and data that's prefetched and ready to be sent over the wire. But agreed that we can reduce the concept count by 1 by just using the command line instead of `ModuleSpec` (see prev comment reply)

The benefit of not treating local compilation differently is that we have "one way of doing things", so better testing. As noted in the commit message, there's no observable overhead to the unnecessary copying.

File: compiler_opt/rl/compilation_runner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mtrofin) [reply compiler_opt/rl/compilation_runner.py:27]: Can we find the pylint blurb that handles that - "google3" doesn't mean anything to the community.

File: compiler_opt/rl/corpus.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yundiqian) [reply compiler_opt/rl/corpus.py:334]: ic, then i think it's fine to not treat local scenario separately.

I think it makes sense to reduce the compilation unit class to 1 (probably re-use the ModuleSpec class) cause it then puts all the compilation-unit operations in one place and could improve the readability a lot.

[the remark this answers is not in the document]
Comment (mtrofin) [reply compiler_opt/rl/corpus.py:334]: can't reduce to 1: we can't economically prefetch all files eagerly. Using 1 class that's modal - i.e. where its state changes depending on function calls - would be equivalent concept count (the object is in 2 states: "not loaded" or "loaded"), but mushed into one type, which would be poor-er separation of concerns.

[the remark this answers is not in the document]
Comment (yundiqian) [reply compiler_opt/rl/corpus.py:334]: ic, let's combine into 2: ModuleSpec and LoadedModuleSpec, where the only difference is that whether the .bc and .thinlto.bc files are loaded. 

and also can you also add the usage example in the comments section at the top of this file? This can help the readability a lot

File: compiler_opt/rl/corpus.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mtrofin) [reply compiler_opt/rl/corpus.py]: Resolving because later discussion tackles this.

File: compiler_opt/rl/corpus.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mtrofin) [reply compiler_opt/rl/corpus.py]: Done - added usage example under `Corpus`, seems like most appropriate.

[the remark this answers is not in the document]
Comment (mtrofin) [reply compiler_opt/rl/corpus.py]: (also rebased)

File: compiler_opt/rl/compilation_runner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mtrofin) [reply compiler_opt/rl/compilation_runner.py]: (and fixed, of course - we should dig around and find how to fix pylint tho, meant that as a "todo")

File: compiler_opt/rl/corpus.py [hunk 6]
```diff
@@ -12,31 +12,145 @@
...
+  """Encapsulates the loaded data of a module and the rules to persist it.
+
+  A LoadedModuleSpec can be passed to a remote location. There, given a local
+  directory, to_module_spec can be called, resulting in the data being saved
+  under that directory, the final compiler command line fully computed, and a
+  ready-to-use FullyQualifiedCmdLine returned.
+  """
+  name: str
+  module: bytes
+  thinlto_index: Optional[bytes] = None
```
Comment (yundiqian) [inline compiler_opt/rl/corpus.py:90]: nit:

module->loaded_ir
thinlto_index->loaded_thinlto_index

to emphasize that they are loaded bytes instead of paths

File: compiler_opt/rl/corpus.py [hunk 7]
```diff
@@ -104,194 +259,161 @@ def __init__(self,
...
+          ret = tuple(f.read().split('\0'))
+          # The options read from a .cmd file must be run with -cc1
+          if ret[0] != '-cc1':
+            raise ValueError('-cc1 flag not present in .cmd file.')
+          return ret
+
+    contents = [
+        ModuleSpec(
+            name=name,
+            size=os.path.getsize(os.path.join(data_path, name + '.bc')),
```
Comment (yundiqian) [inline compiler_opt/rl/corpus.py:311]: how does it work for a trainer on borg case? it currently assumes that the corpus is co-located with the trainer on the same machine (os.path.getsize assumes that, and the repeated bytes reading from .bc and .thinlto.bc files is slow, not sure whether it can finish in the prefetching time)

File: compiler_opt/rl/corpus.py [hunk 8]
```diff
@@ -104,194 +259,161 @@ def __init__(self,
...
+    sampled_specs = self._sampler(self._contents, k=k)
     if sort:
       sampled_specs.sort(key=lambda m: m.size, reverse=True)
     return sampled_specs
 
-  def filter(self, p: re.Pattern):
-    """Filters module specs, keeping those which match the provided pattern."""
-    return Corpus.from_module_specs(
-        (ms for ms in self._module_specs if p.match(ms.name)))
+  def load_module_spec(self, module_spec: ModuleSpec) -> LoadedModuleSpec:
```
Comment (yundiqian) [inline compiler_opt/rl/corpus.py:363]: Looks like should be the init of LoadModuleSpec class?

File: compiler_opt/rl/corpus.py [hunk 9]
```diff
@@ -104,194 +259,161 @@ def __init__(self,
...
+        thinlto_index=thinlto_bytes,
+        orig_options=module_spec.command_line,
+        additional_flags=self._additional_flags,
+        delete_flags=self._delete_flags,
+        replace_flags=self._replace_flags)
 
   @property
-  def module_specs(self):
-    return self._module_specs
+  def contents(self):
```
Comment (yundiqian) [inline compiler_opt/rl/corpus.py:383]: nit: module_specs sounds more informative than contents

File: compiler_opt/rl/corpus.py (hunk 7 above)
Comment (yundiqian) [reply compiler_opt/rl/corpus.py:311]: you can change the os.path.getsize to some tf.gfile api, but it's going to be very slow (and the get_cmdline also becomes slow) so the code here needs to be multi-processed

File: compiler_opt/rl/corpus.py [hunk 10]
```diff
@@ -12,31 +12,145 @@
...
+import tensorflow as tf
 
 from compiler_opt.rl import constant
 
+# Alias to better self-document APIs. Represents a complete, ready to use
+# command line, where all the flags reference existing, local files.
+FullyQualifiedCmdLine = Tuple[str, ...]
+
+
+def _localize_cmdline(
```
Comment (yundiqian) [inline compiler_opt/rl/corpus.py:36]: this function is doing more than localizing cmdline. From readability perspective, how about:
1. still keep the cmd pre-processing(add/delete/replace) in ModuleSpec
2. only takes care of flag localizing here to go over all flags to replace the path prefix.

the benefit is that all the cmd pre-processing code (cmd loading, add/delete/replace conflict detection and merging, actually doing the job) are in one place so that easier to test/debug

File: compiler_opt/rl/corpus.py (hunk 6 above)
Comment (mtrofin) [reply compiler_opt/rl/corpus.py:90]: good point, done

File: compiler_opt/rl/corpus.py (hunk 7 above)
Comment (mtrofin) [reply compiler_opt/rl/corpus.py:311]: (for non-Google contributors, 'borg' refers to distributed case)

ack - left a TODO to keep from bloating the patch. leaving it as `os.path.getsize` so if we do forget to update this, it fails fast (rather than taking a ton of time and still work)

Comment (mtrofin) [reply compiler_opt/rl/corpus.py:311]: nevermind, actually easy to do in this :)

Comment (mtrofin) [reply compiler_opt/rl/corpus.py:311]: done

File: compiler_opt/rl/corpus.py (hunk 8 above)
Comment (mtrofin) [reply compiler_opt/rl/corpus.py:363]: The idea is to keel LoadedModuleSpec just a plain data object, decoupled from provenance. Later, if we want to (for example) fetch corpus from a database, we just need to change how we make LoadedModuleSpecs.

File: compiler_opt/rl/corpus.py (hunk 9 above)
Comment (mtrofin) [reply compiler_opt/rl/corpus.py:383]: done

File: compiler_opt/rl/corpus.py (hunk 10 above)
Comment (mtrofin) [reply compiler_opt/rl/corpus.py]: I like the outcome - even testing is easier - I think the answer to my question above is "yes", see the latest patch.

File: compiler_opt/rl/corpus.py (hunk 8 above)
Comment (yundiqian) [reply compiler_opt/rl/corpus.py:367]: i don't see a lot of benefit in keeping LoadModuleSpec a plain data object, and it already has some functions in it, making it as init will help make the API cleaner. But I'm fine with the current setup :)

File: compiler_opt/rl/corpus.py (hunk 10 above)
Comment (yundiqian) [reply compiler_opt/rl/corpus.py]: yeah, that's what i mean

**Review by yundiqian — APPROVED** [review]
lgtm
